README
== Intro ==

serial-latency-test measures the roundtrip time of packets sent over
the serial port. It calculates the worst case roundtrip time of all
sent packets and displays a histogram of the rountrip time jitter.

== Compile, install and run ==

See https://github.com/koppi/serial-latency-test/wiki

Run serial-latency-test as follows:

 $ serial-latency-test -p /dev/ttyUSB0 -b 9600

   This runs the benchmark on the given serial port with the given baud rate.
   Note, that the RX and TX have to be connected using a cable in the real
   hardware to loop the sent packets back.

== Authors ==

This is an early release. Please report bugs to the authors.

 * Jakob Flierl <jakob.flierl@gmail.com>



debian/changelog
serial-latency-test (0.0.1-2) unstable; urgency=low

  * Switch to dpkg-source 3.0 (git) format
  * Bump version

 -- Jakob Flierl <jakob.flierl@gmail.com>  Sun, 02 Feb 2014 00:17:11 +0100

serial-latency-test (0.0.1-1) unstable; urgency=low

  * Initial release.

 -- Jakob Flierl <jakob.flierl@gmail.com>  Wed, 29 Jan 2014 18:15:50 +0100


src/serial-latency-test.c
/* Serial Latency Tester
 *
 * (C) 2013  Jakob Flierl <jakob.flierl@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>

#include <math.h>

#ifndef INFINITY
#define INFINITY (1.0/0.0)
#endif 

#include <limits.h>
#include <float.h>
#include <signal.h>

#if defined (HAVE_SCHED_H)
#include <sched.h>
#endif

#include <getopt.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>

#if defined (HAVE_SYS_UTSNAME_H)
#include <sys/utsname.h>
#endif

#include "hr_timer.h"

#if defined (HAVE_LINUX_SERIAL_H)
#include <linux/serial.h>
#endif

#include "serial.h"

#define DEBUG 1

#define HISTLEN  10
#define TERMWIDTH 50

#ifndef SQUARE
#define SQUARE(a) ( (a) * (a) )
#endif

#ifndef MIN
#define MIN(a,b) ( (a) < (b) ? (a) : (b) )
#endif

#ifndef MAX
#define MAX(a,b) ( (a) > (b) ? (a) : (b) )
#endif

#ifndef RAIL
#define RAIL(v, min, max) (MIN((max), MAX((min), (v))))
#endif

static int printinterval = 1;

static volatile sig_atomic_t signal_received = 0;

static void fatal(const char *msg, ...)
{
    va_list ap;

    va_start(ap, msg);
    vfprintf(stderr, msg, ap);
    va_end(ap);
    fprintf(stderr, ". Exiting.\n\n");
    exit(EXIT_FAILURE);
}

/* memory allocation error handling */
static void check_mem(void *p)
{
    if (!p)
        fatal("out of memory");
}

#if defined (HAVE_SYS_UTSNAME_H)
void print_uname()
{
    struct utsname u;
    uname (&u);
    printf ("> running on %s release %s (version %s) on %s\n",
            u.sysname, u.release, u.version, u.machine);
}
#endif

/* sets the process to "policy" policy at given priority */
static int set_realtime_priority(int policy, int prio)
{
#if defined (HAVE_SCHED_H)
    struct sched_param schp;
    memset(&schp, 0, sizeof(schp));

    schp.sched_priority = prio;
    if (sched_setscheduler(0, policy, &schp) != 0) {
        perror("sched_setscheduler");
        return -1;
    }
#endif

    return 0;
}

static void usage(const char *argv0)
{
    printf("Usage: %s -p <port> ...\n\n"
           "  -p, --port=port    serial port to run tests on\n"
           "  -b, --baud=baud    baud rate (default: 9600)\n"
#if defined (HAVE_SCHED_H)
           "  -R, --realtime     use realtime scheduling (default: no)\n"
           "  -P, --priority=n   scheduling priority, use with -R\n"
           "                     (default: 99)\n\n"
#endif
           "  -S, --samples=n    to take for the measurement (default: 10000)\n"
           "  -c, --count=n      number of bytes to send per sample (default: 1)\n"
           "  -w, --wait=ms      time interval between measurements (default: 0)\n"
           "  -r, --random-wait  use random interval between wait and 2*wait\n\n"
#if defined (HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
           "  -a, --async        set ASYNC_LOW_LATENCY flag (default: no)\n"
#endif
           "  -x  --xmit=n       set xmit_fifo_size to given number (default: 0)\n"
#endif
           "  -o, --output=file  write the output to file\n\n"
           "  -h, --help         this help\n"
           "  -V, --version      print current version\n\n"
           "Report bugs to Jakob Flierl <jakob.flierl@gmail.com>\n"
           "Website and manual: https://github.com/koppi/serial-latency-test\n"
           "\n", argv0);
}

static void print_version(void)
{
    printf("%s version %s\n", PACKAGE, VERSION);
}

#if defined (_WIN32)
static void wait_ms(double t)
{
    Sleep(t);
}
#else
static void wait_ms(double t)
{
    struct timespec ts;

    ts.tv_sec = t / 1000;
    ts.tv_nsec = (t - ts.tv_sec * 1000) * 1000000;
    nanosleep(&ts, NULL);
}
#endif

/* many thanks to Randall Munroe (http://xkcd.com/221/) */
static int getRandomNumber(void)
{
    return 4; // chosen by fair dice roll.
              // guaranteed to be random.
}

static void sighandler(int sig)
{
	if (!signal_received)
		fprintf(stderr,"\n\n> caught signal - shutting down.\n");
    signal_received = 1;
}

typedef struct {
    PORTTYPE fd;
    int  baud;
    char port[PATH_MAX];
#if defined (HAVE_TERMIOS_H)
    struct termios opts;
#endif
} serial_t;

int digits(double number) {
    int digits = 1, pten = 10;

    while (pten <= number) {
        digits++; pten *= 10;
    }

    return digits;
}

int main(int argc, char *argv[])
{
    setvbuf(stdout, NULL, _IONBF, 0);
    setvbuf(stderr, NULL, _IONBF, 0);

    static struct option long_options[] = {
        {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, 'V'},
#if defined (HAVE_SCHED_H)
        {"realtime", no_argument, NULL, 'R'},
        {"priority", required_argument, NULL, 'P'},
#endif
        {"port", required_argument, NULL, 'p'},
        {"baud", required_argument, NULL, 'b'},
        {"samples", required_argument, NULL, 'S'},
        {"count", required_argument, NULL, 'c'},
        {"wait", required_argument, NULL, 'w'},
        {"random-wait", no_argument, NULL, 'r'},
#if defined (HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
        {"async", no_argument, NULL, 'a'},
#endif
        {"xmit", required_argument, NULL, 'x'},
#endif
        {"output", required_argument, NULL, 'o'},
        {}
    };

#if defined (HAVE_SCHED_H)
    int do_realtime = 0;
    int rt_prio = sched_get_priority_max(SCHED_FIFO);
#endif
#if defined (HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
    int async_low_latency = 0;
#endif
    int xmit_fifo_size = 0;
#endif
    int nr_samples = 10000;
    int nr_count = 1;
    int random_wait = 0;
    double wait = 0.0;
    char output[PATH_MAX];

    serial_t s;

    s.fd   = 0;
    s.baud = 9600;
    snprintf(s.port, sizeof s.port, "%s", "");

    snprintf(output, sizeof output, "%s", "");

    int c;

    while ((c = getopt_long(argc, argv,
                            "h"   /* help */
                            "V"   /* version */
#if defined (HAVE_SCHED_H)
                            "R"   /* realtime */
                            "P:"  /* priority */
#endif
                            "p:"  /* port */
                            "b:"  /* baud */
                            "S:"  /* samples */
                            "c:"  /* count */
                            "w:"  /* wait */
                            "r"   /* random-wait */
#if defined (HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
                            "a"   /* async */
#endif
                            "x:"  /* xmit */
#endif
                            "o:", /* output */
                            long_options, NULL)) != -1) {
        switch (c) {
        case 'h':
            usage(argv[0]);
            return EXIT_SUCCESS;
        case 'V':
            print_version();
            return EXIT_SUCCESS;
#if defined (HAVE_SCHED_H)
        case 'R':
            do_realtime = 1;
            break;
#endif
        case 'p':
            strncpy(s.port, optarg, sizeof(s.port));
            break;
        case 'b':
            s.baud = strtol(optarg, NULL, 10);
            break;
#if defined (HAVE_SCHED_H)
        case 'P':
            rt_prio = atoi(optarg);
            if (rt_prio > sched_get_priority_max(SCHED_FIFO)) {
                printf("> Warning: Given priority:   %d > sched_get_priority_max(SCHED_FIFO)! ", rt_prio);
                printf("Setting priority to %d.\n", sched_get_priority_max(SCHED_FIFO));
                rt_prio = sched_get_priority_max(SCHED_FIFO);
            } else if (rt_prio < sched_get_priority_min(SCHED_FIFO)) {
                printf("> Warning: Given priority:   %d < sched_get_priority_min(SCHED_FIFO)! ", rt_prio);
                printf("Setting priority to %d.\n", sched_get_priority_min(SCHED_FIFO));
                rt_prio = sched_get_priority_min(SCHED_FIFO);
            }
            break;
#endif
        case 'S':
            nr_samples = atoi(optarg);
            if (nr_samples <= 0) {
                printf("> Warning: Given number of samples to take is less or equal zero! ");
                printf("Setting nr of samples to take to 1.\n");
                nr_samples = 1;
            }
            break;
        case 'c':
            nr_count = atoi(optarg);
            break;
        case 'w':
            wait = atof(optarg);
            if (wait < 0) {
                printf("> Warning: Wait time is negative; using zero.\n");
                wait = 0;
            }
            break;
#if defined(HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
        case 'a':
            async_low_latency = 1;
            break;
#endif
        case 'x':
            xmit_fifo_size = atoi(optarg);
            break;
#endif
        case 'r':
            random_wait = 1;
            break;
        case 'o':
            strncpy(output, optarg, sizeof(output));
            break;
        default:
            usage(argv[0]);
            return EXIT_FAILURE;
        }
    }

    if (argc == 1 || argv[optind]) {
        usage(argv[0]);
        return EXIT_FAILURE;
    }

    printf("> ");
    print_version();

#if defined (HAVE_SYS_UTSNAME_H)
    print_uname();
#endif

    if (random_wait)
        srand(getRandomNumber());

#if defined (HAVE_SCHED_H)
    if (do_realtime) {
        printf("> set_realtime_priority(SCHED_FIFO, %d).. ", rt_prio);
        set_realtime_priority(SCHED_FIFO, rt_prio);
        printf("done.\n");
    }
#endif

#if defined (HAVE_TERMIOS_H)
    s.fd = serial_open(s.port, s.baud, &s.opts);
#else
    s.fd = serial_open(s.port, s.baud);
#endif

    if (!s.fd) {
        fatal("Unable to open %s", s.port);
    }

#if defined (HAVE_LINUX_SERIAL_H)
#if defined (ASYNC_LOW_LATENCY)
    if (async_low_latency) {
        if (serial_set_low_latency(s.fd) < 0) {
            fatal("Unable to set ASYNC_LOW_LATENCY on %s", s.port);
        } else {
            printf("> set flag ASYNC_LOW_LATENCY to %d on %s\n", async_low_latency, s.port);
        }
    }
#endif

    if (xmit_fifo_size > 0) {
        if (serial_set_xmit_fifo_size(s.fd, xmit_fifo_size) < 0) {
            fatal("Unable to set xmit_fifo_size %d on %s", xmit_fifo_size, s.port);
        } else {
            printf("> set xmit_fifo_size to %d on %s\n", serial_get_xmit_fifo_size(s.fd), s.port);
        }
    }
#endif

    timerStruct begin, end;

    printf("\n> sampling %d latency values - please wait..\n", nr_samples);
    printf("   event     curr      min      max      avg [ms]\n");

    signal(SIGINT,  sighandler);
    signal(SIGTERM, sighandler);

    double *delays = calloc(nr_samples + 1, sizeof *delays);
    check_mem(delays);

    int cnt_a;
    double min_a, max_a;
    double avg_a;
    double var_m, var_s;

    unsigned int *history = calloc(HISTLEN, sizeof(unsigned int));
    unsigned int histsize = 0;
    double bin_width = 0;
    double bin_min = DBL_MIN;
    unsigned int *histogram = NULL;

    cnt_a = 0;
    min_a = DBL_MAX;
    max_a = 0;
    avg_a = 0;
    var_m = var_s = 0;

    uint8_t *buf_rx = calloc(nr_count + 1, sizeof (uint8_t));
    uint8_t *buf_tx = calloc(nr_count + 1, sizeof (uint8_t));

    unsigned int i;

    for (i = 0; i < nr_count; ++i) {
        buf_tx[i] = i % 255;
    }

    time_t last = time(NULL);

    int err = 0;

    for (c = 0; c < nr_samples; ++c) {
        if (wait) {
            if (random_wait)
                wait_ms(wait + rand() * wait / RAND_MAX);
            else
                wait_ms(wait);
            if (signal_received)
                break;
        }

        int n = 0;

        GetHighResolutionTime(&begin);

        n = serial_write(s.fd, buf_tx, nr_count);

        if (n != nr_count) {
            fprintf(stderr, "serial_write() n = %d nr_count = %d\n", n, nr_count);
            err = 1;
            signal_received = 1;
        }

        if (signal_received)
            break;

        int rd = 0;

        while(!rd) {
            n = serial_read(s.fd, buf_rx, nr_count); // blocking read using select

            if (n == nr_count) {
                rd = 1;
            } else {
                rd = 1;
                signal_received = 1;
                err = 1;
                fprintf(stderr, "serial_read() n = %d nr_count = %d\n", n, nr_count);
            }
        }

        if (signal_received)
            break;

        GetHighResolutionTime(&end);

        double delay = ConvertTimeDifferenceToSec(&end, &begin) * 1000.0;

        delays[cnt_a] = delay;

        time_t now = time(NULL);

        if (printinterval > 0 && now >= last + printinterval) {
            last = now;
            if (cnt_a > 0)
                printf("\n");
        }

        avg_a += delay;

        if (delay < min_a) min_a = delay;
        if (delay > max_a) max_a = delay;

        printf(" %7d %8.2f %8.2f %8.2f %8.2f\r", cnt_a, delay, min_a, max_a, avg_a / (double)cnt_a);

        /* histogram */
        if (cnt_a < HISTLEN) {
            history[cnt_a] = delay;
        } else if (cnt_a == HISTLEN) {
            int j;
            double stddev = 0;
            const double avg = avg_a / (double)HISTLEN;
            for (j = 0; j < HISTLEN; ++j) {
                stddev += SQUARE((double)history[j] - avg);
            }
            stddev = sqrt(stddev/(double)HISTLEN);
            // Scott's normal reference rule
            bin_width = 3.5 * stddev * pow(HISTLEN, -1.0/3.0);
            int k = ceil((double)(max_a - min_a) / bin_width);

            bin_min = min_a;
            if (bin_min > bin_width) { k++; bin_min -= bin_width; }
            if (bin_min > bin_width) { k++; bin_min -= bin_width; }
            histsize = k+2;

/*
  if (printinterval > 0) {
  printf("\n -- initializing histogram with %d bins (min:%.2f w:%.2f [samples]) --\n", histsize, bin_min, bin_width);
  }
*/
            histogram = calloc(histsize + 1,sizeof(unsigned int));
            for (j = 0; j < HISTLEN; ++j) {
                int bin = RAIL(floor(((double)history[j] - bin_min) / bin_width), 0, histsize);
                histogram[bin]++;
            }
        } else {
            int bin = RAIL(floor(((double)delay - bin_min) / bin_width), 0, histsize);
            histogram[bin]++;
        }

        cnt_a++;
    }

    if (strlen(output)) {
        FILE *fp = fopen(output, "w");

        if (!fp) {
            fatal("unable to open output file '%s'", output);
        }

        for (i = 0; i < cnt_a; ++i) {
            fprintf(fp, "%8.2f\n", delays[i]);
        }

        fclose(fp);
    }

    if (!err) {
        printf("\n> done.\n\n");
    } else {
        printf("\n> done (with errors).\n\n");
    }

    if (histsize > 0) {
        printf("> latency distribution:\n\n");
        int i,j;
        int binlevel = 0;
        for (i = 0; i < histsize; ++i) {
            if (histogram[i] > binlevel) binlevel = histogram[i];
        }

        if (binlevel > 0) {
            int dig = digits(max_a); char fmt[256];
            snprintf(fmt, sizeof(fmt), " %%%d.2f .. %%%d.2f [ms]: %%%dd ", dig + 4, dig + 4, digits(cnt_a));
            for (i = 0; i <= histsize; ++i) {
                double hmin, hmax;
                if (i == 0) {
                    hmin = 0.0;
                    hmax = bin_min;
                } else if (i == histsize) {
                    hmin = bin_min + (double)(i-1) * bin_width;
                    hmax = INFINITY;
                } else {
                    hmin = bin_min + (double)(i-1) * bin_width;
                    hmax = bin_min + (double)(i) * bin_width;
                }
                printf(fmt, hmin, hmax, histogram[i]);
                int bar_width = (histogram[i] * TERMWIDTH ) / binlevel;
                if (bar_width == 0 && histogram[i] > 0) bar_width = 1;
                for (j = 0; j < bar_width; ++j) printf("#");
                printf("\n");
            }
        }

        printf("\n");
        printf(" best    latency was %.2f ms\n", min_a);
        printf(" worst   latency was %.2f ms\n", max_a);
        printf(" average latency was %.2f ms\n", avg_a / (double)cnt_a);
        printf("\n");
    }

    free(histogram);
    free(history);

    free(delays);
    free(buf_rx);
    free(buf_tx);

#if defined(HAVE_TERMIOS_H)
    serial_close(s.fd, &s.opts);
#else
    serial_close(s.fd);
#endif

    return EXIT_SUCCESS;
}


autogen.sh
#! /bin/sh

if [ "$USER" = "root" ]; then
	echo "*** You cannot do this as "$USER" please use a normal user account."
	exit 1
fi
if test ! -f configure.ac ; then
	echo "*** Please invoke this script from directory containing configure.ac."
	exit 1
fi

echo "running aclocal"
aclocal
rc=$?

if test $rc -eq 0; then
	echo "running libtool"
	libtoolize --force --automake --copy
	rc=$?
else
	echo "An error occured, autogen.sh stopping."
	exit $rc
fi

if test $rc -eq 0; then
	echo "libtool worked."
else
	echo "libtool not found. trying glibtool."
	glibtoolize --force --automake --copy
	rc=$?
fi

if test $rc -eq 0; then
        echo "running autoheader"
        autoheader
        rc=$?
else
        echo "An error occured, autogen.sh stopping."
        exit $rc
fi

if test $rc -eq 0; then
        echo "running autoconf"
        autoconf
        rc=$?
else
        echo "An error occured, autogen.sh stopping."
        exit $rc
fi

if test $rc -eq 0; then
	echo "running automake"
	automake -f -i --add-missing --copy
	rc=$?
else
	echo "An error occured, autogen.sh stopping."
	exit $rc
fi

echo "autogen.sh complete"
exit $rc


src/serial.h
#ifndef SERIAL_H
#define SERIAL_H

#ifdef __cplusplus
extern "C" {
#endif

#if defined (HAVE_CONFIG_H)
#include "config.h"
#endif

#if defined (HAVE_TERMIOS_H)
#include <termios.h>
#else
#include <windows.h>
#endif

#include <unistd.h>
#include <stdint.h>

#if defined (HAVE_TERMIOS_H)
#define PORTTYPE int
#else
#define PORTTYPE HANDLE
#endif

#if defined (HAVE_TERMIOS_H)
	PORTTYPE serial_open(const char *port, int baud, struct termios *opts);
	int		 serial_close(PORTTYPE fd, struct termios *opts);
#if defined (HAVE_LINUX_SERIAL_H)
	int      serial_set_xmit_fifo_size(PORTTYPE fd, int size);
	int      serial_get_xmit_fifo_size(PORTTYPE fd);
#if defined (ASYNC_LOW_LATENCY)
	int      serial_set_low_latency(PORTTYPE fd);
#endif
#endif
#else
	PORTTYPE serial_open(const char *port, int baud);
	int		 serial_close(PORTTYPE fd);
#endif
	ssize_t	 serial_writebyte(PORTTYPE fd, uint8_t byte);
	ssize_t	 serial_write(PORTTYPE fd, const uint8_t *buf, size_t len);
	ssize_t	 serial_read(PORTTYPE fd, uint8_t *buf, size_t len);

#ifdef __cplusplus
} /* extern "C" */
#endif

#endif


src/serial.c
#include "serial.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#if defined (HAVE_TERMIOS_H)
#include <termios.h>
#endif
#if defined (HAVE_SYS_IOCTL_H)
#include <sys/ioctl.h>
#endif
#if defined (HAVE_LINUX_SERIAL_H)
#include <linux/serial.h>
#endif
#if defined (HAVE_LINUX_IOCTL_H)
#include <linux/ioctl.h>
#endif
#if defined (HAVE_ASM_IOCTLS_H)
#include <asm/ioctls.h>
#endif

// #define DBG(M, ...)
#define DBG(M, ...) fprintf(stderr, "%s:%d: " M "\n", __FILE__, __LINE__, ##__VA_ARGS__)

#define clean_errno() (errno == 0 ? "None" : strerror(errno))
#define log_err(M, ...) fprintf(stderr, "%s:%d: errno: %s " M "\n", __FILE__, __LINE__, clean_errno(), ##__VA_ARGS__)

ssize_t serial_writebyte(PORTTYPE fd, uint8_t byte)
{
	return serial_write(fd, &byte, 1);
}

ssize_t serial_write(PORTTYPE fd, const uint8_t *buf, size_t len)
{
#if defined (_WIN32)
	DWORD n;
	BOOL r;
	r = WriteFile(fd, buf, len, &n, NULL);
	if (!r) return 0;
	return n;
#else
	ssize_t n = write(fd, buf, len);

	if (n != len) {
		DBG("len=%ld, n=%ld, error=%d %s", len, n, errno, strerror(errno));
	}

	if (n != len) return -1;

	return n;
#endif
}

ssize_t serial_read(PORTTYPE fd, uint8_t *buf, size_t len)
{
	ssize_t count = 0;

#if defined (_WIN32)
	COMMTIMEOUTS timeout;
	DWORD n;
	BOOL r;
	int waiting=0;

	GetCommTimeouts(fd, &timeout);
	timeout.ReadIntervalTimeout = MAXDWORD; // non-blocking
	timeout.ReadTotalTimeoutMultiplier = 0;
	timeout.ReadTotalTimeoutConstant = 0;
	SetCommTimeouts(fd, &timeout);
	while (count < len) {
		r = ReadFile(fd, buf + count, len - count, &n, NULL);
		if (!r) {
			log_err("read error r=%d count=%ld len=%ld", r, count, len);
			return 0;
		}
		if (n > 0) count += n;
		else {
			if (waiting) break;	 // 1 sec timeout
			timeout.ReadIntervalTimeout = MAXDWORD;
			timeout.ReadTotalTimeoutMultiplier = MAXDWORD;
			timeout.ReadTotalTimeoutConstant = 1000;
			SetCommTimeouts(fd, &timeout);
			waiting = 1;
		}
	}

#else
	int r;
	int retry = 0;

	/*
	if (len > sizeof(buf) || len < 1) {
		fprintf(stderr, "serial_read() len=%d sizeof(buf)=%ld\n", len, sizeof(buf));
		return -1;
		}*/

	// non-blocking read mode
	//fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK);

	while (count < len) {
		r = read(fd, buf + count, len - count);
		//printf("read, r = %d\n", r);
		if (r < 0 && errno != EAGAIN && errno != EINTR) return -1;
		else if (r > 0) count += r;
		else {
			// no data available right now, must wait
			fd_set fds;
			struct timeval t;
			FD_ZERO(&fds);
			FD_SET(fd, &fds);
			t.tv_sec = 1;
			t.tv_usec = 0;
			r = select(fd+1, &fds, NULL, NULL, &t);
			// printf("select, r = %d\n", r);
			if (r < 0) return -1;
			if (r == 0) return count; // timeout
		}
		retry++;
		if (retry > 10000) return -100; // no input
	}
	//fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) & ~O_NONBLOCK);
#endif

	if (count != len) {
		DBG("len=%ld, n=%ld, error=%d %s", len, count, errno, strerror(errno));
	}

	return count;
}

#if defined (HAVE_TERMIOS_H)
PORTTYPE serial_open(const char *port, int baud, struct termios *opts)
#else
	PORTTYPE serial_open(const char *port, int baud)
#endif
{
	PORTTYPE fd;

	DBG("opening port %s @ %d bps", port, baud);

	if (!strlen(port)) {
		log_err("invalid port name %s", port);
		return 0;
	}

#if defined (_WIN32)
	COMMCONFIG cfg;
	COMMTIMEOUTS timeout;
	DWORD n;
	char portname[256];
	int num;
	if (sscanf(port, "COM%d", &num) == 1) {
		sprintf(portname, "\\\\.\\COM%d", num); // Microsoft KB115831
	} else {
		strncpy(portname, port, sizeof(portname)-1);
		portname[n-1] = 0;
	}
	fd = CreateFile(portname, GENERIC_READ | GENERIC_WRITE,
					0, 0, OPEN_EXISTING, 0, NULL);
	if (fd == INVALID_HANDLE_VALUE) {
		log_err("unable to open port %s", port);
		return 0;
	}

	GetCommConfig(fd, &cfg, &n);

	cfg.dcb.BaudRate = baud;
	cfg.dcb.fBinary = TRUE;
	cfg.dcb.fParity = FALSE;
	cfg.dcb.fOutxCtsFlow = FALSE;
	cfg.dcb.fOutxDsrFlow = FALSE;
	cfg.dcb.fOutX = FALSE;
	cfg.dcb.fInX = FALSE;
	cfg.dcb.fErrorChar = FALSE;
	cfg.dcb.fNull = FALSE;
	cfg.dcb.fRtsControl = RTS_CONTROL_ENABLE;
	cfg.dcb.fAbortOnError = FALSE;
	cfg.dcb.ByteSize = 8;
	cfg.dcb.Parity = NOPARITY;
	cfg.dcb.StopBits = ONESTOPBIT;
	cfg.dcb.fDtrControl = DTR_CONTROL_ENABLE;
	SetCommConfig(fd, &cfg, n);
	GetCommTimeouts(fd, &timeout);
	timeout.ReadIntervalTimeout = 0;
	timeout.ReadTotalTimeoutMultiplier = 0;
	timeout.ReadTotalTimeoutConstant = 1000;
	timeout.WriteTotalTimeoutConstant = 0;
	timeout.WriteTotalTimeoutMultiplier = 0;
	SetCommTimeouts(fd, &timeout);

	return fd;
#else
	//fd = open(port, O_RDWR | O_NOCTTY);
	fd = open(port, O_RDWR | O_NOCTTY | O_NONBLOCK);

	if (fd == -1)  {
		log_err("open() failed");
		return 0;
	}

#if defined (HAVE_TERMIOS_H)
	if (tcgetattr(fd, opts) < 0) {
		log_err("tcgetattr() failed");
		return 0;
	}

	struct termios toptions;

	if (tcgetattr(fd, &toptions) < 0) {
		log_err("tcgetattr()");
		return 0;
	}

	speed_t brate = baud; // let you override switch below if needed

	switch(baud) {
#define B(x) case x: brate=B##x; break;
		B(50);		B(75);		B(110);		B(134);		B(150);
		B(200);		B(300);		B(600);		B(1200);	B(1800);
		B(2400);	B(4800);	B(9600);	B(19200);	B(38400);
		B(57600);	B(115200);	B(230400);	B(460800);
#if defined B500000
		B(500000); B(576000); B(921600); B(1000000); B(1152000);
		B(1500000); B(2000000); B(2500000);B(3000000); B(3500000); B(4000000);
#endif

#undef B
	default:
		log_err("unknown baud rate %d", baud);
		return 0;
		break;
	}

	cfsetispeed(&toptions, brate);
	cfsetospeed(&toptions, brate);

	// 8N1
	toptions.c_cflag &= ~PARENB;
	toptions.c_cflag &= ~CSTOPB;
	toptions.c_cflag &= ~CSIZE;
	toptions.c_cflag |= CS8;

	// no flow control
	toptions.c_cflag &= ~CRTSCTS;

	// turn off s/w flow ctrl
	toptions.c_iflag &= ~(IXON | IXOFF | IXANY);

	// turn on READ & ignore ctrl lines
	toptions.c_cflag |= CREAD | CLOCAL;

	// make raw
	toptions.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
	toptions.c_oflag &= ~OPOST;

	// see: http://unixwiz.net/techtips/termios-vmin-vtime.html
	toptions.c_cc[VMIN]	 = 0;
	toptions.c_cc[VTIME] = 10;

	if (tcsetattr(fd, TCSANOW, &toptions) < 0) {
		log_err("tcsetattr() failed");
		return 0;
	}

	int flags = fcntl(fd, F_GETFL);
	if (fcntl(fd, F_SETFL, flags & (~O_NONBLOCK))) {
		log_err("fcntl() failed");
		return 0;
	}

	if (tcflush(fd, TCIOFLUSH) < 0) {
		log_err("tcflush() failed");
		return 0;
	}
#endif

#endif
	return fd;
}

#if defined (HAVE_TERMIOS_H)
int serial_close(PORTTYPE fd, struct termios *opts)
#else
	int serial_close(PORTTYPE fd)
#endif
{

#if defined (HAVE_TERMIOS_H)
  if (tcsetattr(fd, TCSANOW, opts) < 0) {
	  log_err("tcsetattr() failed");
  }
#endif

#if defined (_WIN32)
	CloseHandle(fd);
#else
	return close(fd);
#endif
}

#if defined (ASYNC_LOW_LATENCY)
int serial_set_low_latency(PORTTYPE fd) {
	struct serial_struct ser_info;

	if (ioctl(fd, TIOCGSERIAL, &ser_info) < 0) {
		log_err("ioctl(TIOCGSERIAL) failed");
		return -1;
	}

	ser_info.flags |= ASYNC_LOW_LATENCY;

	if (ioctl(fd, TIOCSSERIAL, &ser_info) < 0) {
		log_err("ioctl(TIOCSSERIAL) failed");
	} else {
		DBG("set ASYNC_LOW_LATENCY mode.");
	}

	return 0;
}
#endif

#if defined (HAVE_LINUX_SERIAL_H)
int serial_set_xmit_fifo_size(PORTTYPE fd, int size) {
	struct serial_struct ser_info;

	if (ioctl(fd, TIOCGSERIAL, &ser_info) < 0) {
		log_err("ioctl(TIOCGSERIAL) failed");
		return -1;
	}

	DBG("xmit_fifo_size was %d\n", ser_info.xmit_fifo_size);

	ser_info.xmit_fifo_size = size;

	if (ioctl(fd, TIOCSSERIAL, &ser_info) < 0) {
		log_err("ioctl(TIOCSSERIAL) failed");
		return -1;
	}

	return 0;
}

int serial_get_xmit_fifo_size(PORTTYPE fd) {
	struct serial_struct ser_info;

	if (ioctl(fd, TIOCGSERIAL, &ser_info) < 0) {
		log_err("ioctl(TIOCGSERIAL) failed");
		return -1;
	}

	return ser_info.xmit_fifo_size;
}
#endif
